### PR TITLE
feat: add native support for get_json_object expression

### LINKED ISSUE #3162: [Feature] Support Spark expression: get_json_object
## What is the problem the feature request solves?

> **Note:** This issue was generated with AI assistance. The specification details have been extracted from Spark documentation and may need verification.

Comet does not currently support the Spark `get_json_object` function, causing queries using this function to fall back to Spark's JVM execution instead of running natively on DataFusion.

The `GetJsonObject` expression extracts JSON values from a JSON string using JSONPath expressions. It takes a JSON string and a JSONPath query, returning the matching value(s) as a string representation.

Supporting this expression would allow more Spark workloads to benefit from Comet's native acceleration.

### Spark Specification

**Syntax:**
```sql
get_json_object(json_string, path)
```

```scala
// DataFrame API usage
col("json_column").getItem(path)
// or using expr()
expr("get_json_object(json_column, '$.field')")
```

**Arguments:**
| Argument | Type | Description |
|----------|------|-------------|
| json | Expression | The JSON string to query against |
| path | Expression | The JSONPath expression to extract values |

**Return Type:** Returns `StringType` - the extracted JSON value as a string representation.

**Supported Data Types:**
- **json parameter**: String type containing valid JSON
- **path parameter**: String type containing valid JSONPath expressions

**Edge Cases:**
- **Null handling**: Returns null if either json or path parameters are null
- **Invalid JSON**: Returns null for malformed JSON input strings
- **Invalid JSONPath**: Returns null for syntactically incorrect JSONPath expressions
- **No matches**: Returns null when the JSONPath doesn't match any elements
- **Multiple matches**: For array results, returns JSON array string representation

**Examples:**
```sql
-- Extract simple field
SELECT get_json_object('{"name":"John","age":30}', '$.name');
-- Result: "John"

-- Extract from array
SELECT get_json_object('[{"a":"b"},{"a":"c"}]', '$[*].a');
-- Result: ["b","c"]

-- Extract nested field
SELECT get_json_object('{"user":{"profile":{"name":"Alice"}}}', '$.user.profile.name');
-- Result: "Alice"

-- Array element access
SELECT get_json_object('{"items":["apple","banana","cherry"]}', '$.items[1]');
-- Result: "banana"
```

```scala
// DataFrame API usage
import org.apache.spark.sql.functions._

df.select(get_json_object(col("json_data"), "$.field"))

// Using expr for complex JSONPath
df.select(expr("get_json_object(json_column, '$[*].nested.field')"))

// Multiple extractions
df.select(
  get_json_object(col("json_data"), "$.name").alias("name"),
  get_json_object(col("json_data"), "$.age").alias("age")
)
```

### Implementation Approach

See the [Comet guide on adding new expressions](https://datafusion.apache.org/comet/contributor-guide/adding_a_new_expression.html) for detailed instructions.

1. **Scala Serde**: Add expression handler in `spark/src/main/scala/org/apache/comet/serde/`
2. **Register**: Add to appropriate map in `QueryPlanSerde.scala`
3. **Protobuf**: Add message type in `native/proto/src/proto/expr.proto` if needed
4. **Rust**: Implement in `native/spark-expr/src/` (check if DataFusion has built-in support first)

## Additional context

**Difficulty:** Large
**Spark Expression Class:** `org.apache.spark.sql.catalyst.expressions.GetJsonObject`

**Related:**
- `json_tuple` - Extract multiple JSON fields in a single operation
- `from_json` - Parse JSON string into structured data types
- `to_json` - Convert structured data to JSON strings
- `json_array_length` - Get length of JSON arrays

---
*This issue was auto-generated from Spark reference documentation.*

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes #3162.

## Rationale for this change

`get_json_object` is a widely-used Spark function for extracting values from JSON strings using JSONPath expressions. Without native support, queries using this function fall back to Spark's JVM execution. This PR adds an initial native implementation to allow Comet to accelerate these queries.

This is a starting point. The expression is marked `Incompatible` and is **disabled by default**. Users must set `spark.comet.expression.GetJsonObject.allowIncompatible=true` to enable it.

## What changes are included in this PR?

**Rust implementation** (`native/spark-expr/src/string_funcs/get_json_object.rs`):
- Custom JSONPath parser supporting `$` (root), `.field`, `['field']` (bracket notation), `[n]` (array index), and `[*]` (array wildcard)
- Path evaluation with separate fast-path for non-wildcard paths (zero Vec allocations) and wildcard paths
- Uses `serde_json` with `preserve_order` feature for Spark-compatible key ordering
- 19 unit tests

**Scala serde** (`spark/src/main/scala/org/apache/comet/serde/strings.scala`):
- `CometGetJsonObject` with `getSupportLevel` returning `Incompatible` (Spark's Jackson parser allows single-quoted JSON and unescaped control characters that `serde_json` does not)

**Registration and wiring:**
- Added to `stringExpressions` map in `QueryPlanSerde.scala`
- Registered in `comet_scalar_funcs.rs` via `scalarFunctionExprToProtoWithReturnType`

**SQL tests** (`get_json_object.sql`): 30 test queries covering field extraction, nested objects, arrays, wildcards, nulls, invalid JSON, bracket notation, edge cases.

**Docs**: Updated `expressions.md` and `spark_expressions_support.md`.

### Current performance

Benchmarked with 1M rows of JSON (~200 bytes each) on Apple M3 Ultra:

| Case | Spark (ms) | Comet (ms) | Relative |
|------|-----------|------------|----------|
| Simple field (`$.name`) | 705 | 785 | 0.9X |
| Numeric field (`$.age`) | 725 | 789 | 0.9X |
| Nested field (`$.address.city`) | 773 | 805 | 1.0X |
| Array element (`$.items[0]`) | 734 | 795 | 0.9X |
| Nested object (`$.address`) | 869 | 926 | 0.9X |

Comet is currently ~10% slower than Spark. The primary reason is that `serde_json` parses the full JSON document into a DOM tree on every row, while Spark's Jackson-based implementation uses a streaming parser that can skip irrelevant fields without allocating.

### Known limitations and future work

This is an initial implementation. Known gaps that could be addressed in follow-up PRs:

1. **Streaming JSON parser**: Replace `serde_json::from_str` (full DOM parse) with a streaming approach (e.g., `jiter` or custom `serde_json::Deserializer` with `IgnoredAny`) to skip irrelevant JSON content without allocating. This would likely close the performance gap with Spark.
2. **`$.*` on arrays**: Spark distinguishes `$.*` (object wildcard, using `Wildcard` token) from `$[*]` (array wildcard, using `Subscript::Wildcard`). Our parser treats both as the same `Wildcard` segment. Currently `$.*` on arrays returns values in Comet but null in Spark.
3. **Double wildcard flattening**: Spark's `$[*][*]` triggers `FlattenStyle` which flattens nested arrays. Our implementation doesn't handle this special case.
4. **Single wildcard match after index**: For patterns like `$.arr[0][*].field`, Spark's `WriteStyle` state machine may produce different wrapping behavior than our count-based approach.
5. **`preserve_order` is workspace-wide**: Cargo unifies features, so enabling `preserve_order` on `serde_json` in `spark-expr` also enables it for all other crates in the workspace. Could be addressed by isolating the JSON parsing behind a feature flag.

## How are these changes tested?

- 19 Rust unit tests covering path parsing and evaluation edge cases
- 30 SQL-file-based tests (`CometSqlFileTestSuite`) that run each query through both Spark and Comet and compare results, with dictionary encoding on/off
- Microbenchmark (`CometGetJsonObjectBenchmark`) comparing Spark vs Comet performance across 5 query patterns

### CODE REVIEW

File: native/spark-expr/src/string_funcs/get_json_object.rs [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+                        None => builder.append_null(),
+                    }
+                }
+            }
+
+            Ok(ColumnarValue::Array(Arc::new(builder.finish())))
+        }
+        // Scalar json, scalar path
+        (ColumnarValue::Scalar(json_scalar), ColumnarValue::Scalar(path_scalar)) => {
+            let json_str = match json_scalar {
```
Comment (comphead) [inline native/spark-expr/src/string_funcs/get_json_object.rs]: maybe it is good to have an utility method to deduplicate code around `json_str` and `path_str`

**Review by comphead — APPROVED** [review]
Thanks @andygrove from my first review, I think this is awesome and code looks like to unicode safe, however to prove it, would be nice having come unicode tests

File: native/spark-expr/src/string_funcs/get_json_object.rs (hunk 1 above)
Comment (andygrove) [reply native/spark-expr/src/string_funcs/get_json_object.rs]: Updated

**Comment by andygrove** [comment]
> Thanks @andygrove from my first review, I think this is awesome and code looks like to unicode safe, however to prove it, would be nice having come unicode tests

Thanks. I added unicode tests
